kroxylicious-runtime/src/main/java/io/kroxylicious/proxy/internal/util/ConnectionRegistry.java

```diff
@@ -0,0 +1,263 @@
+/*
+ * Copyright Kroxylicious Authors.
+ *
+ * Licensed under the Apache Software License version 2.0, available at http://www.apache.org/licenses/LICENSE-2.0
+ */
+
+package io.kroxylicious.proxy.internal.util;
+
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.atomic.AtomicInteger;
+
+import io.micrometer.core.instrument.Gauge;
+
+import edu.umd.cs.findbugs.annotations.NonNull;
+import edu.umd.cs.findbugs.annotations.Nullable;
+
+/**
+ * A thread-safe registry for tracking active connections in the Kroxylicious proxy.
+ *
+ * This singleton class maintains counts of active connections between clients and proxy,
+ * and between proxy and servers. It automatically creates and manages Micrometer gauges
+ * to expose these connection counts as metrics.
+ *
+ * The registry tracks connections per virtual cluster and node ID combination,
+ * using "bootstrap" as the node ID for bootstrap connections.
+ */
+public final class ConnectionRegistry {
+
+    /** Singleton instance of the connection registry. */
+    private static final ConnectionRegistry INSTANCE = new ConnectionRegistry();
+
+    /** Map storing active connection counts from clients to proxy, keyed by cluster:nodeId. */
+    private final ConcurrentHashMap<String, AtomicInteger> clientToProxyActiveConnections = new ConcurrentHashMap<>();
+
+    /** Map storing active connection counts from proxy to servers, keyed by cluster:nodeId. */
+    private final ConcurrentHashMap<String, AtomicInteger> proxyToServerActiveConnections = new ConcurrentHashMap<>();
+
+    /** Map storing Micrometer gauges for client-to-proxy connections, keyed by cluster:nodeId. */
+    private final ConcurrentHashMap<String, Gauge> clientToProxyGauges = new ConcurrentHashMap<>();
+
+    /** Map storing Micrometer gauges for proxy-to-server connections, keyed by cluster:nodeId. */
+    private final ConcurrentHashMap<String, Gauge> proxyToServerGauges = new ConcurrentHashMap<>();
+
+    /**
+     * Private constructor to enforce singleton pattern.
+     */
+    private ConnectionRegistry() {
+    }
+
+    /**
+     * Increments the count of active connections from clients to the proxy.
+     *
+     * This method is thread-safe and will automatically create a Micrometer gauge
+     * for the specified cluster and node if one doesn't already exist.
+     *
+     * @param clusterName the name of the virtual cluster
+     * @param nodeId the node ID, or null for bootstrap connections
+     */
+    public static void incrementClientToProxyConnections(@NonNull String clusterName, @Nullable Integer nodeId) {
```

Comment:
incrementClientToProxyActiveConnections? Just to avoid confusion with other connection metrics?

Reply:
N/A now. Removed ConnectionRegistry.